src/hazardous/secret_stream/xchacha20poly1305.rs

```diff
@@ -0,0 +1,455 @@
+	pub fn new(key: chacha20Key, nonce: Nonce) -> Self {
+		let mut s = SecretStreamXChaCha20Poly1305 {
+			key: hchacha20(&key, &nonce.as_ref()[..HCHACHA_NONCESIZE]).unwrap(),
+			nonce: [0u8; IETF_CHACHA_NONCESIZE],
+		};
+		s.reset_counter();
+		s.get_inonce_mut()
+			.copy_from_slice(&nonce.as_ref()[HCHACHA_NONCESIZE..]);
+		s
+	}
+
+	/// prepare a new push or pull operation
+	pub fn init(&mut self, key: chacha20Key, nonce: Nonce) {
+		*self = SecretStreamXChaCha20Poly1305::new(key, nonce);
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// derives a new secret key used for encryption/decryption
+	pub fn rekey(&mut self) -> Result<(), UnknownCryptoError> {
+		let mut new_key_and_inonce =
+			[0u8; CHACHA_KEYSIZE + SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES]; //TODO
+
+		new_key_and_inonce[..CHACHA_KEYSIZE].copy_from_slice(&self.key[..CHACHA_KEYSIZE]);
+		new_key_and_inonce[CHACHA_KEYSIZE..]
+			.copy_from_slice(&self.nonce[SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES..]);
+
+		chacha20_enc_in_place(
+			&self.get_key(),
+			&self.get_nonce(),
+			0,
+			&mut new_key_and_inonce,
+		)?;
+
+		self.key[..CHACHA_KEYSIZE].copy_from_slice(&new_key_and_inonce[..CHACHA_KEYSIZE]);
+		self.nonce[SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES..]
+			.copy_from_slice(&new_key_and_inonce[CHACHA_KEYSIZE..]);
+
+		self.reset_counter();
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// //Ecryptes a message
+	pub fn push(
+		&mut self,
+		plaintext: &[u8],
+		ad: Option<&[u8]>,
+		dst_out: &mut [u8],
+		tag: Tag,
+	) -> Result<(), UnknownCryptoError> {
+		const BLOCKSIZE: usize = 64;
+		let mut block = [0u8; BLOCKSIZE];
+		let mut slen = [0u8; 8];
+		let pad = [0u8; 16];
+		let adlen = match ad {
+			Some(v) => v.len(),
+			None => 0,
+		};
+		let msglen = plaintext.len();
+		let cipherpos = core::mem::size_of::<Tag>();
+		let macpos = cipherpos + msglen;
+		if dst_out.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES + msglen {
+			return Err(UnknownCryptoError);
+		}
+
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 0, &mut block)?;
+		let mut poly = init(&OneTimeKey::from_slice(&block[..POLY1305_KEYSIZE])?);
+		use zeroize::Zeroize;
+		block.zeroize();
+		if adlen > 0 {
+			poly.update(ad.unwrap())?;
+		}
+		poly.update(&pad[..((16usize.overflowing_sub(adlen).0) & 15)])?;
+		block[0] = tag.bits();
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 1, &mut block)?;
+		poly.update(&block)?;
+		dst_out[0] = block[0];
+
+		chacha20_enc(
+			&self.get_key(),
+			&self.get_nonce(),
+			2,
+			plaintext,
+			&mut dst_out[cipherpos..],
+		)?;
+		poly.update(&dst_out[cipherpos..(cipherpos + msglen)])?;
+		poly.update(
+			&pad[..((16usize
+				.overflowing_sub(BLOCKSIZE)
+				.0
+				.overflowing_add(msglen)
+				.0) & 15)],
+		)?;
+		store_u64_into_le(&[adlen as u64], &mut slen);
+		poly.update(&slen)?;
```

Comment:
We can split up the authentication into a separate function, like it has been done in `chacha20poly1305::process_authentication`, to avoid code-duplication and make it easier to review.